sound/soc/sof/ipc.c

```diff
@@ -672,9 +672,43 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
 	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
 	struct sof_ipc_fw_version *v = &ready->version;
 	struct sof_ipc_ctrl_data_params sparams;
+	struct snd_sof_widget *swidget;
+	bool widget_found = false;
 	size_t send_bytes;
 	int err;
 
+	list_for_each_entry_reverse(swidget, &sdev->widget_list, list) {
+		if (swidget->comp_id == scontrol->comp_id) {
+			widget_found = true;
+			break;
+		}
+	}
+
+	if (!widget_found) {
+		dev_err(sdev->dev, "error: can't find widget with id %d\n", scontrol->comp_id);
+		return -EINVAL;
+	}
+
+	/*
+	 * return 0 for SET commands as the cached value will be sent to the DSP when the widget
+	 * has been set up. Return error for GET to notify the userspace that the widget
+	 * data is inaccessible.
+	 */
+	if (!atomic_read(&swidget->use_count)) {
```

Comment:
what race are you hinting it. Assuming, this gets preempted because the pipeline was just started, even in that case, the value being cached will be sent after the widgets is set up in the process no?

Reply:
Oh get on the other hand  if premempted could return INVALID after the widget has been setup. But even if were to protect the use_count increment with a lock, a get would have returned -EINVAL isnt it? Do you have a suggestion for what we can do here?

Reply:
@ranj063 I'm also wondering about the use of atomics. This does not seem like very performance critical section (done when opening/closing PCMs), so a simple lock would be easier to review. But I wonder if existing framework level locks are already taken here, or is this really a case where we have sharing that is not covered by existing locking. Is the case here to open multiple PCMs, which share the same widgets?

The use count is used in multiple places (in later patches), so reviewing correctness in case of concurrent access, is not trivial.

Reply:
@kv2019i for PCM open/close the framelevel locks app;y, so may not even need atomic there. But there's no lock that covers both control IO and PCM operations. So maybe I can add a widget use_count_mutex instead of using atomic? What do you think?